system/src/system_power_manager.cpp

```diff
@@ -552,52 +555,205 @@ void PowerManager::handleStateChange(battery_state_t from, battery_state_t to, b
+      clearIntermadiateBatteryState();
+      handleCharging();
+    } else {
+      // We've passed the jitter window and the battery is charged indeed
+      DBG_PWR("Charged!");
+      confirmBatteryState(g_batteryState, BATTERY_STATE_CHARGED);
+      // Restore normal recharge threshold
+      DBG_PWR("Restore normal threshold: 100mV");
+      power.setRechargeThreshold(PMIC_NORMAL_RECHARGE_THRESHOLD); // 100mV
+    }
+    return;
+  }
+
+  if (millis() - batMonitorTimeStamp_ >= battMonitorPeriod_) {
+    batMonitorTimeStamp_ = millis();
+    if (config_.flags & HAL_POWER_CHARGE_STATE_DISABLE) {
+      deduceBatteryStateChargeDisabled();
+    } else {
+      deduceBatteryStateChargeEnabled();
+    }
   }
 }
 
-void PowerManager::checkWatchdog() {
-  if (g_batteryState == BATTERY_STATE_DISCONNECTED &&
-      ((millis() - chargingDisabledTimestamp_) >= DEFAULT_WATCHDOG_TIMEOUT)) {
-    // Re-enable charging, do not run DPDM detection
-    LOG_DEBUG(TRACE, "re-enabling charging");
-    chargingDisabledTimestamp_ = 0;
-    g_batteryState = BATTERY_STATE_UNKNOWN;
-    initDefault(false);
+void PowerManager::deduceBatteryStateChargeDisabled() {
+  DBG_PWR("deduceBatteryStateChargeDisabled()");
+  handleCharging();
+  // XXX: This is a dirty hack. Using the VCELL to deduce the battery state
+  // is based on the experience that when charging is disabled, the VCELL can
+  // correctly be measured.
+  // When charging is disabled, the VCELL can reliably reflect the voltage on VBAT
+  FuelGauge fuel(true);
+  float vCell = fuel.getVCell();
+  DBG_PWR("VCell: %dmV", (uint16_t)(vCell * 1000));
+
+  // Make an assumption
+  if (g_batteryState == BATTERY_STATE_UNKNOWN) {
+    DBG_PWR("Battery state unknown, make an assumption:");
+    if (vCell < BATTERY_CONNECTED_VCELL_THRESHOLD) {
+      confirmBatteryState(g_batteryState, BATTERY_STATE_DISCONNECTED);
+    } else {
+      confirmBatteryState(g_batteryState, BATTERY_STATE_NOT_CHARGING);
+    }
+    return;
+  } else if (g_batteryState == BATTERY_STATE_DISCONNECTED) {
+    if (vCell > BATTERY_CONNECTED_VCELL_THRESHOLD) {
+      if (vcellDebounceCount_ == 0) {
+        // It's probably that the battery is attached, request to shorten the monitor period
+        battMonitorPeriod_ = BATTERY_STATE_CHANGE_CHECK_PERIOD;
+        DBG_PWR("Battery might be attached");
+      }
+      vcellDebounceCount_++;
+      if (vcellDebounceCount_ >= BATTERY_VCELL_DEBOUNCE_COUNT) {
+        goto connected;
+      }
+    } else {
+      confirmBatteryState(g_batteryState, BATTERY_STATE_DISCONNECTED);
+    }
+    return;
+  }
+
+connected:
+  PMIC power(true);
+  const uint8_t status = power.getSystemStatus();
+  const uint8_t pwrGood = (status >> 2) & 0b01;
+  if ((status & 0x08) || !pwrGood) {
+    DBG_PWR("In DPM mode or power is not good");
+    // It's in DPM mode, the battery is discharging
+    confirmBatteryState(g_batteryState, BATTERY_STATE_DISCHARGING);
```

Comment:
I think this is a bit confusing when battery charging is enabled and we have a power source connected. We will be reporting a sticky `discharging` state even though we are still charging the battery while at the same time occasionally taking some charge off of it to compensate for spikes. I would probably stick with just checking `pwrGood` here.

cc @eberseth